### PR TITLE
update nce-sm

### PR DESCRIPTION
**Pull request by Victor0118**
@HTAustin @tuzhucheng @rosequ  Could you review my code when available?

I add noise constraintive estimation (NCE) to the standard SM model. #10 
The results on TrecQA are as follows:

##### Without NCE

Metric |rand   |static|non-static|multichannel
-------|-------|------|----------|------------
MAP    |0.7441 |0.7524|0.7688    |0.7641
MRR    |0.8172 |0.8012|0.8144    |0.8174

##### Max Neg Sample

To be added

##### Pairwise + Random Sample and neg_num = 8

Metric |rand   |static|non-static|multichannel
-------|-------|------|----------|------------
MAP    |0.7427 |0.7546|0.7716    |0.7794
MRR    |0.8151 |0.8061|0.8347    |0.8467

This is my first pull request to Castor. Please be careful. Thanks!

### CODE REVIEW

File: nce/NCE-Pairwise-SM/train.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+        elif distance == "cosine":
+            # feat_norm = normalize(np.array(neg_dict[key]["feature"].reshape(-1,1)), norm='l2')
+            neg_feature = np.array(neg_dict[key]["feature"])
+            feat_norm = neg_feature / np.sqrt(sum(neg_feature ** 2))
+            dis = 1 - feat_norm.dot(pos_feature_norm)
+        dis_list.append(dis)
+        neg_list.append(key)
+
+    k = min(k, len(neg_dict))
+    min_list = heapq.nsmallest(k, enumerate(dis_list), key=operator.itemgetter(1))
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/train.py]: You can merge `neg_list` and `dis_list` into `neg_dis_dict` and sort this dict by value to get top k id.

File: nce/NCE-Pairwise-SM/train.py [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+        return []
+    k = min(k, len(q2neg[qid_i]))
+    ran = random.sample(q2neg[qid_i], k)
+    return ran
+
+# pack the lists of question/answer/ext_feat into a torchtext batch
+def get_batch(question, answer, ext_feat, size):
+    new_batch = data.Batch()
+    new_batch.batch_size = size
+    new_batch.dataset = batch.dataset
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/train.py]: Where do you define the `batch`?

File: nce/NCE-Pairwise-SM/train.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    setattr(new_batch, "answer", torch.stack(answer))
+    setattr(new_batch, "question", torch.stack(question))
+    setattr(new_batch, "ext_feat", torch.stack(ext_feat))
+    return new_batch
+
+
+
+while True:
+    if early_stop:
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/train.py]: What is the point of setting `early_stop` here? It doesn't even run any epoch.

File: nce/NCE-Pairwise-SM/train.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+    new_batch.batch_size = size
+    new_batch.dataset = batch.dataset
+    setattr(new_batch, "answer", torch.stack(answer))
+    setattr(new_batch, "question", torch.stack(question))
+    setattr(new_batch, "ext_feat", torch.stack(ext_feat))
+    return new_batch
+
+
+
+while True:
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/train.py]: `While True` is not good. Could you set a maximum num of epoches and add `while epoch < max_epoches`

File: nce/NCE-Pairwise-SM/train.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+                question2answer[qid_i]["pos"][aid_i]["ext_feat"] = ext_feat_i
+
+                # get neg samples in the first epoch but do not train
+                if epoch == 1:
+                    continue
+                # random generate sample in the first training epoch
+                elif epoch != 1:
+                    near_list = get_random_neg_id(q2neg, qid_i, k=args.neg_num)
+                else:
+                    debug_qid = qid_i
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/train.py]: remove this

File: nce/NCE-Pairwise-SM/train.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+                neg_batch = get_batch(new_train_neg["question"], new_train_neg["answer"], new_train_neg["ext_feat"],
+                                      true_batch_size)
+
+                optimizer.zero_grad()
+                output = pw_model([pos_batch, neg_batch])
+
+                '''
+                debug code
+                '''
+                cmp = output[:, 0] <= output[:, 1]
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/train.py]: For the debug and evaluation code, could you add a class for them. So we can evaluate the loss, evaluation metrics easily for different sets.

File: nce/NCE-Pairwise-SM/create_dataset.sh [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (rosequ) [reply nce/NCE-Pairwise-SM/create_dataset.sh]: @HTAustin this is just a preprocessing step so that the data can be easily split into fields in torchtext's `data.TabularDataset`

File: nce/NCE-Pairwise-SM/main.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+random.seed(args.seed)
+
+QID = data.Field(sequential=False)
+AID = data.Field(sequential=False)
+QUESTION = data.Field(batch_first=True)
+ANSWER = data.Field(batch_first=True)
+LABEL = data.Field(sequential=False)
+EXTERNAL = data.Field(sequential=False, tensor_type=torch.FloatTensor, batch_first=True, use_vocab=False,
+                      preprocessing=data.Pipeline(lambda x: x.split()),
+                      postprocessing=data.Pipeline(lambda x, train: [float(y) for y in x]))
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/main.py]: @Victor0118 with the latest torchtext version, you'll have to replace the code with following:
```python
EXTERNAL = data.Field(sequential=True, tensor_type=torch.FloatTensor, batch_first=True, use_vocab=False,
            postprocessing=data.Pipeline(lambda arr, _, train: [float(y) for y in arr]))
```

File: nce/NCE-Pairwise-SM/main.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    exit()
+
+QID.build_vocab(train, dev, test)
+AID.build_vocab(train, dev, test)
+QUESTION.build_vocab(train, dev, test)
+ANSWER.build_vocab(train, dev, test)
+LABEL.build_vocab(train, dev, test)
+
+train_iter = data.Iterator(train, batch_size=args.batch_size, device=args.gpu, train=True, repeat=False,
+                                   sort=False, shuffle=True)
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/main.py]: `sort` is None by default. So you can drop the argument. Check [source](https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L76)

File: nce/NCE-Pairwise-SM/model.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        # append external features and feed to fc
+        x.append(x_ext)
+        x = torch.cat(x, 1)
+        x = F.tanh(self.combined_feature_vector(x))
+
+        '''
+        add dropout and hidden layer here? No, it should be in PairwiseConv
+        '''
+        # x = self.dropout(x)
+        # x = self.hidden(x)
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/model.py]: unused comments

File: nce/NCE-Pairwise-SM/overlap_features.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+from collections import defaultdict
+from argparse import ArgumentParser
+
+from nltk.stem.porter import PorterStemmer
+
+def load_data(dname):
+    stemmer = PorterStemmer()
+    qids, questions, answers, labels = [], [], [], []
+    print('Load folder ' + dname)
+    with open(dname+'a.toks', encoding='utf-8') as f:
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/overlap_features.py]: space around `+`

File: nce/NCE-Pairwise-SM/model.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+
+class PairwiseConv(nn.Module):
+    """docstring for PairwiseConv"""
+    def __init__(self, model):
+        super(PairwiseConv, self).__init__()
+        # self.linearLayer = self:LinearLayer() ??
+        # self.convModel = SmPlusPlus(config)
+        self.convModel = model
+        self.dropout = nn.Dropout(0.5)
```
[the remark this answers is not in the document]
Comment (rosequ) [reply nce/NCE-Pairwise-SM/model.py]: Follow [PEP 8] (https://www.python.org/dev/peps/pep-0008/#imports) style imports.

File: nce/NCE-Pairwise-SM/trec_dataset.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+from torchtext import data
```
[the remark this answers is not in the document]
Comment (rosequ) [reply nce/NCE-Pairwise-SM/trec_dataset.py]: Follow PEP 8 import style

File: nce/NCE-Pairwise-SM/overlap_features.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
+import numpy as np
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/overlap_features.py]: You probably don't need this file anymore if you are cool with the overlap calculations in datasets.wikiqa.WIKIQA.

Let me know if otherwise.

File: nce/NCE-Pairwise-SM/evaluate.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import shlex
+import subprocess
+
+def evaluate(instances, valid, config):
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/evaluate.py]: You don't need this anymore, import `utils.relevancy_metrics.get_map_mrr` and use that.

File: nce/NCE-Pairwise-SM/create_dataset.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (tuzhucheng) [reply nce/NCE-Pairwise-SM/create_dataset.sh]: You don't need this anymore, you can just import `datasets.trecqa.TRECQA` and `datasets.wikiqa.WIKIQA` and use those.

You might need to extend those with answer id. I can do it for you if it's easier. Let's talk in-person about this.

File: nce/NCE-Pairwise-SM/trec_dataset.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+import os
+
+class TrecDataset(data.TabularDataset):
+    dirname = 'data'
+    @classmethod
+
+    def splits(cls, question_id, question_field, answer_id, answer_field, external_field, label_field,
+               train='train.tsv', validation='dev.tsv', test='test.tsv'):
+        path = './data'
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/trec_dataset.py]: in the new version of torchtext, the splits method is updated:
```python 
def splits(cls, question_id, question_field, answer_id, answer_field, external_field, label_field, root='.data',
               train='trecqa.train.tsv', validation='trecqa.dev.tsv', test='trecqa.test.tsv'):
        path = './data'
        return super(TrecDataset, cls).splits(
            path, root, train, validation, test,...
```

File: nce/NCE-Pairwise-SM/trec_dataset.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tuzhucheng) [reply nce/NCE-Pairwise-SM/trec_dataset.py]: Check out `datasets.{trecqa,wikiqa}`.

File: nce/NCE-Pairwise-SM/create_dataset.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (rosequ) [reply nce/NCE-Pairwise-SM/create_dataset.sh]: @tuzhucheng I want to discuss this too. For all my experiments, I follow a similar pipeline and I may not be able to use your script as I have more fields. Maybe, we can update your script to have different fields...

**Comment by HTAustin** [comment]
The current version is quite messy and quite hard to follow. We might need to refactor it in the near future. The current code is in fact running on random neg examples which is not correct. @Victor0118  I think running a correct experiment now is the most important thing to do.

File: nce/NCE-Pairwise-SM/trec_dataset.py [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+        prefix_name = 'trecqa.'
+        return super(TrecDataset, cls).splits(
+            os.path.join(path, prefix_name), train, validation, test,
```
Comment (HTAustin) [inline nce/NCE-Pairwise-SM/trec_dataset.py]: `os.path.join(path, prefix_name), train, validation, test` will return `./data/trecqa./train.tsv`

Comment (rosequ) [reply nce/NCE-Pairwise-SM/trec_dataset.py]: @HTAustin that's because you're using the new torchtext. @Victor0118 will have to update his code to suit the new torchtext. Also, @Victor0118 update your pytorch to the recent version and enable `torch.backends.cudnn.deterministic = True`

File: nce/NCE-Pairwise-SM/create_dataset.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/create_dataset.sh]: @tuzhucheng @rosequ  Yes. I also have more field. We can add on @tuzhucheng 's code, which is shared by everybody.

File: nce/NCE-Pairwise-SM/train.py [hunk 18]
```diff
@@ -0,0 +1,420 @@
...
+parameter = filter(lambda p: p.requires_grad, pw_model.parameters())
+
+# the SM model originally follows SGD but Adadelta is used here
+optimizer = torch.optim.Adadelta(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=1e-6)
+# A good lr is required to use Adam
+# optimizer = torch.optim.Adam(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=1e-8)
+
+
+# criterion = nn.CrossEntropyLoss()
+marginRankingLoss = nn.MarginRankingLoss(margin = 1, size_average = False)
```
[the remark this answers is not in the document]
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/train.py]: You are right. Previously I set a variable total_sample_per_batch and want to calculate the average loss between two dev logs. Now I think printing loss of one batch for each dev log is already OK.

File: nce/NCE-Pairwise-SM/train.py (hunk 1 above)
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/train.py]: Good suggestion. But the code will look almost the same and heap sort will be a little faster.

File: nce/NCE-Pairwise-SM/overlap_features.py (hunk 11 above)
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/overlap_features.py]: Do you mean I need to add space for every `+`, right?

File: nce/NCE-Pairwise-SM/main.py (hunk 9 above)
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/main.py]: When I delete the argument, I got an error:
`TypeError: '<' not supported between instances of 'Example' and 'Example' `

Does it work well in your environment?

Comment (tuzhucheng) [reply nce/NCE-Pairwise-SM/main.py]: You can look at how I'm using my TrecQA dataset here: https://github.com/castorini/Castor/blob/master/mp_cnn/dataset.py#L51

File: nce/NCE-Pairwise-SM/overlap_features.py (hunk 11 above)
Comment (rosequ) [reply nce/NCE-Pairwise-SM/overlap_features.py]: yes, space around the operators

File: nce/NCE-Pairwise-SM/main.py (hunk 9 above)
Comment (rosequ) [reply nce/NCE-Pairwise-SM/main.py]: @Victor0118 did you try this on the new torchtext? In any case, you can follow Michael's code.

**Comment by Victor0118** [comment]
@HTAustin @tuzhucheng @rosequ 
Please check the updated version. The current number is slightly lower than my previous implementation [here](https://github.com/Victor0118/NCE_SM_Pytorch/tree/original_framework).  I think it is related to the shared data loader. I will check it in the future.

File: nce/NCE-Pairwise-SM/main.py [hunk 19]
```diff
@@ -82,15 +81,9 @@ def predict(test_mode, dataset_iter):
         qid_array = index2qid[np.transpose(dev_batch.qid.cpu().data.numpy())]
         true_label_array = index2label[np.transpose(dev_batch.label.cpu().data.numpy())]
 
-        # scores = model(dev_batch)
         output = model.convModel(dev_batch)
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/main.py]: remove extra newlines

File: nce/NCE-Pairwise-SM/train.py [hunk 20]
```diff
@@ -80,7 +75,7 @@ def set_vectors(field, vector_path):
 train_iter = data.Iterator(train, batch_size=args.batch_size, device=args.gpu, train=True, repeat=False,
                            sort=False, shuffle=True)
 dev_iter = data.Iterator(dev, batch_size=args.batch_size, device=args.gpu, train=False, repeat=False,
-                         sort=False, shuffle=False)
+                         sort=False,    shuffle=False)
```
Comment (rosequ) [inline nce/NCE-Pairwise-SM/train.py]: extra space before shuffle?

File: nce/NCE-Pairwise-SM/trec_dataset.py (hunk 16 above)
Comment (rosequ) [reply nce/NCE-Pairwise-SM/trec_dataset.py]: @Victor0118 did you run this on the latest torchtext version? if so, `root` is one of the args to `splits()` method. Please double check this.

File: datasets/trecqa.py [hunk 21]
```diff
@@ -26,29 +31,63 @@ def __init__(self, path):
...
+            for i, token in enumerate(field.vocab.itos):
+                wv_index = stoi.get(token, None)
+                if wv_index is not None:
+                    field.vocab.vectors[i] = vectors[wv_index]
+                else:
+                    # initialize <unk> with U(-0.25, 0.25) vectors
+                    field.vocab.vectors[i] = torch.FloatTensor(dim).uniform_(-0.05, 0.05)
+        else:
+            print("Error: Need word embedding pt file")
+            print("Error: Need word embedding pt file")
```
Comment (rosequ) [inline datasets/trecqa.py]:
```python
print("Error: Need word embedding pt file")
``` 
is repeated twice

File: datasets/castor_dataset.py [hunk 22]
```diff
@@ -19,7 +20,7 @@ class CastorPairDataset(Dataset, metaclass=ABCMeta):
     LABEL_FIELD = None
 
     @abstractmethod
-    def __init__(self, path):
+    def __init__(self, path, additional_fields=None, examples_extra=None, load_ext_feats = False):
```
Comment (tuzhucheng) [inline datasets/castor_dataset.py]: Please don't use any spaces in default values for keyword args per PEP 8 specifications.

`load_ext_feats = False` => `load_ext_feats=False`

File: datasets/trecqa.py [hunk 23]
```diff
@@ -26,29 +31,63 @@ def __init__(self, path):
         """
         Create a TRECQA dataset instance
         """
-        super(TRECQA, self).__init__(path)
+        aid_list = []
+        fields = [("aid", self.AID_FIELD)]
+        with open(os.path.join(path, 'ans_id.txt'), 'r') as aid_file:
```
Comment (tuzhucheng) [inline datasets/trecqa.py]: How and where is this file (`ans_id.txt`) generated?

File: datasets/trecqa.py [hunk 24]
```diff
@@ -14,10 +15,14 @@ class TRECQA(CastorPairDataset):
     NAME = 'trecqa'
     NUM_CLASSES = 2
     ID_FIELD = Field(sequential=False, tensor_type=torch.FloatTensor, use_vocab=False, batch_first=True)
+    AID_FIELD = Field(sequential=False, use_vocab=False, batch_first=True)
     TEXT_FIELD = Field(batch_first=True, tokenize=lambda x: x)  # tokenizer is identity since we already tokenized it to compute external features
-    EXT_FEATS_FIELD = Field(tensor_type=torch.FloatTensor, use_vocab=False, batch_first=True, tokenize=lambda x: x)
+    EXT_FEATS_FIELD = Field(tensor_type=torch.FloatTensor, use_vocab=False, batch_first=True, tokenize=lambda x: x,
+                            postprocessing=Pipeline(lambda arr, _, train: [float(y) for y in arr]))
```
Comment (tuzhucheng) [inline datasets/trecqa.py]: Did you try setting `sequential` to `False`. I'm wondering maybe if you do that then you no longer need to `postprocessing` argument.

See this: https://github.com/pytorch/text/blob/v0.2.0/torchtext/data/field.py#L294. If `sequential` is false then torchtext seems to do the conversion for you.

File: datasets/trecqa.py [hunk 25]
```diff
@@ -14,10 +15,14 @@ class TRECQA(CastorPairDataset):
...
     LABEL_FIELD = Field(sequential=False, use_vocab=False, batch_first=True)
 
+    VOCAB_NUM = 0
```
Comment (tuzhucheng) [inline datasets/trecqa.py]: Maybe `VOCAB_SIZE` is better.

File: datasets/trecqa.py [hunk 26]
```diff
@@ -26,29 +31,63 @@ def __init__(self, path):
...
+            cls.TEXT_FIELD = cls.set_vectors(cls.TEXT_FIELD, os.path.join(vectors_dir, vectors_name))
+
+        cls.LABEL_FIELD.build_vocab(train, validation, test)
 
-        cls.TEXT_FIELD.build_vocab(train, validation, test, vectors=vectors)
+        cls.VOCAB_NUM = len(cls.TEXT_FIELD.vocab)
 
-        return BucketIterator.splits((train, validation, test), batch_size=batch_size, repeat=False, shuffle=shuffle, device=device)
+        # BucketIterator
+        return Iterator.splits((train, validation, test), batch_size=batch_size, repeat=False, shuffle=shuffle, device=device)
```
Comment (tuzhucheng) [inline datasets/trecqa.py]: Why are we changing from `BucketIterator` to `Iterator`?

`BucketIterator` "minimizes amount of padding needed while producing freshly shuffled batches for each new epoch."

File: nce/NCE-Pairwise-SM/README.md [hunk 27]
```diff
@@ -0,0 +1,88 @@
+## NCE-SM model
+
+#### References:
++ Aliaksei _S_everyn and Alessandro _M_oschitti. 2015. Learning to Rank Short Text Pairs with Convolutional Deep Neural
+Networks. In Proceedings of the 38th International ACM SIGIR Conference on Research and Development in Information 
+Retrieval (SIGIR '15). ACM, New York, NY, USA, 373-382. DOI: http://dx.doi.org/10.1145/2766462.2767738
+
++ Jinfeng Rao, Hua He, and Jimmy Lin. [Noise-Contrastive Estimation for Answer Selection with Deep Neural Networks.](http://dl.acm.org/citation.cfm?id=2983872) *Proceedings of the 25th ACM International on Conference on Information and Knowledge Management (CIKM 2016)*, pages 1913-1916.
+
+### Requirements
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/README.md]: Let's get rid of these requirements. They should be specified in `requirements.txt` and the PyTorch version specified here is out of date anyways.

File: nce/NCE-Pairwise-SM/README.md [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+```
+nltk==3.2.2
+numpy==1.11.3
+pytorch==0.1.12
+```
+
+The code uses torchtext for text processing. Set torchtext:
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/README.md]: Update these instructions to install using `pip` from `PyPI` instead of installing from source.

File: nce/NCE-Pairwise-SM/README.md [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+```bash
+git clone https://github.com/pytorch/text.git
+cd text
+python setup.py install
+```
+
+We use `trec_eval` for evaluation:
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/README.md]: This is no longer needed, I have a script that does this.

File: nce/NCE-Pairwise-SM/main.py [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+import torch
+from torchtext import data
+
+from args import get_args
+from trec_dataset import TrecDataset
+from utils.relevancy_metrics import get_map_mrr
+
+from datasets.trecqa import TRECQA
+import os
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/main.py]: Group Python module together. According to [PEP 8](https://www.python.org/dev/peps/pep-0008/):

> Imports should be grouped in the following order:
> standard library imports
> related third party imports
> local application/library specific imports
> You should put a blank line between each group of imports.

File: nce/NCE-Pairwise-SM/main.py [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+
+if not args.cuda:
+    args.gpu = -1
+if torch.cuda.is_available() and args.cuda:
+    logger.info("Note: You are using GPU for training")
+    torch.cuda.set_device(args.gpu)
+    torch.cuda.manual_seed(args.seed)
+if torch.cuda.is_available() and not args.cuda:
+    logger.info("Warning: You have Cuda but do not use it. You are using CPU for training")
+np.random.seed(args.seed)
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/main.py]: Group these seed setting statements with `torch.manual_seed(args.seed)` rather than scattering them at different places.

File: nce/NCE-Pairwise-SM/requirements.txt [hunk 32]
```diff
@@ -0,0 +1,3 @@
+nltk==3.2.1
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/requirements.txt]: This is out of date, please update.

Also add in torchtext.

File: nce/NCE-Pairwise-SM/train.py [hunk 33]
```diff
@@ -0,0 +1,374 @@
...
+import numpy as np
+import random
+
+import torch
+import torch.nn as nn
+from torchtext import data
+
+from args import get_args
+from model import SmPlusPlus, PairwiseConv
+import operator
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/train.py]: Same comment about grouping imports (see earlier comment)

File: nce/NCE-Pairwise-SM/train.py [hunk 34]
```diff
@@ -0,0 +1,374 @@
...
+            print("Shift model to GPU")
+
+        pw_model = PairwiseConv(model)
+
+    parameter = filter(lambda p: p.requires_grad, pw_model.parameters())
+
+    # the SM model originally follows SGD but Adadelta is used here
+    optimizer = torch.optim.Adadelta(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=args.eps)
+    # A good lr is required to use in the following optimizer
+    # optimizer = torch.optim.Adam(parameter, lr=args.lr, weight_decay=args.weight_decay, eps=1e-8)
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/train.py]: Should we make them configurable through command line args instead of commenting them out?

File: nce/NCE-Pairwise-SM/train.py [hunk 35]
```diff
@@ -0,0 +1,374 @@
...
+
+    start = time.time()
+    header = '  Time Epoch Iteration Progress    (%Epoch)  Average_Loss Train_Accuracy Dev/MAP  Dev/MRR'
+    dev_log_template = ' '.join(
+        '{:>6.0f},{:>5.0f},{:>9.0f},{:>5.0f}/{:<5.0f} {:>7.0f}%,{:>11.6f},{:>11.6f},{:12.6f},{:8.4f}'.split(','))
+    log_template = ' '.join('{:>6.0f},{:>5.0f},{:>9.0f},{:>5.0f}/{:<5.0f} {:>7.0f}%,{:>11.6f},{:>11.6f},'.split(','))
+    os.makedirs(args.save_path, exist_ok=True)
+    os.makedirs(os.path.join(args.save_path, args.dataset), exist_ok=True)
+    print(header)
+
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/train.py]: Too many blank lines

File: nce/NCE-Pairwise-SM/train.py [hunk 36]
```diff
@@ -0,0 +1,374 @@
...
+                                                   pt_file=True, device=args.gpu, unk_init=UnknownWordVecCache.unk) #
+
+    index2text = np.array(TRECQA.TEXT_FIELD.vocab.itos)
+
+
+    config.target_class = 2
+    config.questions_num = TRECQA.VOCAB_NUM
+    config.answers_num = TRECQA.VOCAB_NUM
+
+    print("index2text:", index2text)
```
Comment (tuzhucheng) [inline nce/NCE-Pairwise-SM/train.py]: You used a logger before, why use print when you can use logger.

File: nce/NCE-Pairwise-SM/train.py (hunk 34 above)
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/train.py]: Good suggestion! thx.

File: datasets/trecqa.py (hunk 26 above)
Comment (Victor0118) [reply datasets/trecqa.py]: You are right. I use Iterator because NCE will select samples so padding is necessary. Then I suppose during the test stage there should be also padding. I will run an experiment to check it tonight. I will add a bool variable to choose the Iterator/BucketIterator if it is true. Otherwise, I will change to BucketIterator only.

File: datasets/trecqa.py (hunk 24 above)
Comment (Victor0118) [reply datasets/trecqa.py]: I will get an error if I delete the `postprocessing` argument and add `sequential=False`:
```
for batch_idx, batch in enumerate(iter(train_iter)): 
.....
TypeError: only length-1 arrays can be converted to Python scalars
```
I use torchtext v0.2 and pytorch 0.2.0 now.

File: datasets/wikiqa.py [hunk 37]
```diff
@@ -10,7 +10,7 @@
 from datasets.idf_utils import get_pairwise_word_to_doc_freq, get_pairwise_overlap_features
 
 
-class WikiQA(CastorPairDataset):
+class WIKIQA(CastorPairDataset):
```
Comment (tuzhucheng) [inline datasets/wikiqa.py]: I changed this to lower case and it's now changed back haha..

Comment (Victor0118) [reply datasets/wikiqa.py]: oh...Sorry. It is due to my messy local code. I will learn from it. Thanks!

File: nce/NCE-Pairwise-SM/overlap_features.py (hunk 11 above)
Comment (tuzhucheng) [reply nce/NCE-Pairwise-SM/overlap_features.py]: I think you forgot to add space around operator (@rosequ's comment): `dname+'a.toks'` -> `dname + 'a.toks'`

File: nce/NCE-Pairwise-SM/train.py [hunk 38]
```diff
@@ -0,0 +1,420 @@
...
+                if epoch == 1:
+                    continue
+                # random generate sample in the first training epoch
+                elif epoch != 1:
+                    near_list = get_random_neg_id(q2neg, qid_i, k=args.neg_num)
+                else:
+                    debug_qid = qid_i
+                    near_list = get_nearest_neg_id(features[i], question2answer[qid_i]["neg"], distance="cosine", k=args.neg_num)
+
+                batch_near_list.extend(near_list)
```
[the remark this answers is not in the document]
Comment (Victor0118) [reply nce/NCE-Pairwise-SM/train.py]: Yes. I did this in pointwise version. But I found pairwise model involves too many changes in the model. So I have to make it a separate model. Jinfeng also did the same thing with me. Probably in the future, I will consider how to combine them.

File: nce/NCE-Pairwise-SM/trec_dataset.py (hunk 17 above)
Comment (rosequ) [reply nce/NCE-Pairwise-SM/trec_dataset.py]: Yes, I do but I haven't merged this. You can refer the following snippet:
```python
from torchtext import data

class TrecDataset(data.TabularDataset):
    dirname = 'data'
    @classmethod

    def splits(cls, question_id, question_field, answer_field, 
                    external_field, label_field, root='.data',
                    train='trecqa.train.tsv', validation='trecqa.dev.tsv',
                    test='trecqa.test.tsv'):
        path = './data'
        return super(TrecDataset, cls).splits(
            path, root, train, validation, test,
            format='TSV', fields=[('qid', question_id), ('label', label_field),
                                  ('question', question_field), ('ext_feat', external_field),
                                  ('answer', answer_field)])
```
notice the mandatory argument `root`
